docs: improve README (#106)
openshift-service-mesh/federation: README.md

# Federation

> This project implements Istio mesh federation using a Kubernetes controller that provides an API
> for managing multi-mesh communication, implements service discovery and automates the management of Istio configuration.

Reviewer: WDYT about this slight rephrasing? 

```suggestion
This project enables Istio mesh federation through a Kubernetes controller and offers an API for managing multi-mesh communication. It automates service discovery and configuration of Istio.
```

Mesh federation enables mTLS communication between applications across mesh boundaries.
Each mesh can federate a subset of its services to allow applications from other meshes to connect to these services.

Federated services are exposed on a passthrough gateway, so mTLS is not terminated at the edge of the cluster,
and authorization can be performed by the federated application.

## Multi-primary vs federation

[Multi-primary](https://istio.io/latest/docs/setup/install/multicluster/multi-primary_multi-network/) and
[primary-remote](https://istio.io/latest/docs/setup/install/multicluster/primary-remote_multi-network/) topologies
are great solutions for expanding single mesh to multiple k8s clusters for better system resiliency and higher availability.
However, these solutions do not fit well in the following cases:

1. Decentralized control and ownership of clusters.

    **Use case**: Different teams or departments manage their own clusters and control planes independently.

    **Reason**: Federation allows each team to maintain autonomy over their cluster’s Istio configurations while still enabling
    selective cross-mesh communication.

> 1. Simplified networking between clusters.

Reviewer 2: I think it's more about smaller attack surface right? Because we don't need to share API server credentials and don't need to publicly expose the API server to make it work

Author: Not quite. Direct access to kube-apiserver is not insecure per se, because users can enable various security mechanisms, like VPN, firewall, TLS, authn, authz, RBAC etc. But in multi-mesh deployments, environments are independent from each other, and enabling all those mechanisms may not be possible. Suggesting that exposing kube-apiserver is insecure in general, might suggest that single mesh multi-cluster solutions are insecure by design, while it's true.

Controllers DO NOT enforce any authz policy at the mesh boundaries to avoid mTLS termination between applications.
Application or cluster admins are responsible for configuring their authz policies, and it is highly recommended
to deny all traffic by default and allow only selected services.

## Identity and trust model

This controller does not provide any mechanism to share trust bundles between meshes using different CAs.
It can only enable mTLS communication between meshes when all clusters use the same root CA or use SPIRE
with enabled trust bundle federation.

## Getting started

Follow these guides to see how it works in practice:
1. [Simple multi-mesh bookinfo deployment](examples/README.md).
2. [Integration with SPIRE](examples/spire/README.md).

## Comparing to other projects

#### Admiral

[Admiral](https://github.com/istio-ecosystem/admiral) is primarily designed to manage multi-cluster service discovery
and traffic distribution in Istio, focusing on use cases where clusters are part of a single logical mesh,
such as multi-primary or primary-remote topologies, so it does not seem like a right place to implement multi-mesh APIs.

#### Emcee

> [Emcee](https://github.com/istio-ecosystem/emcee) was a PoC of multi-mesh deployment, but that project has been inactive for 5 years,
> and there is no clear value in contributing our ideas, because that would require to change all APIs and general assumptions for that project,
> so it's easier to start from scratch.

Reviewer: Similar reasons as above.

```suggestion
[Emcee](https://github.com/istio-ecosystem/emcee), a proof-of-concept for mesh federation, has been inactive for over five years. Reviving the project would require significant changes to its APIs and underlying assumptions, making it more practical to start fresh rather than build on the existing codebase.
```

> A Kubernetes controller that provides discovery service for Istio mesh federation.

Reviewer 3: As this is the first sentence people will see for the project, I think it could be expanded a bit to introduce the project and describe the controller in more in depth. Perhaps something like (borrowed from your other document):

"This project implements Istio mesh federation using a Kubernetes controller that provides an API for managing multi-mesh communication, implements service discovery and automates the management of Istio configuration."

Author: Thanks, done.

Mesh federation enables mTLS communication between applications across mesh boundaries.
Each mesh can federate a subset of its services to allow applications from other meshes to connect to these services.

Federated services are exposed on a passthrough gateway, so mTLS is not terminated at the edge of the cluster,
and authorization can be performed by the federated application.

## How it works

> Controllers are deployed with sidecars like any other application in the mesh. Controllers connect to remote peers

Reviewer 3: There's a lot described in this paragraph, so expanding some of the sentences to be more descriptive and to include more steps may help to better understand the architecture diagram below. For example...
```suggestion
The federation controller is deployed within each federated mesh with a sidecar like any other application. It uses a local ingress gateway and remote network configuration to connect to remote federation controllers
```
I don't think I have that quite right - please correct, but I just wanted to have the steps broken down a bit more to understand how the federation controller works better.

Author: Thanks, I improved this section. Please look once again.

and send a discovery request to subscribe to federated services.

> When a controller receives discovery response with a federated service, it creates `ServiceEntriy` or `WorkloadEntry`

Reviewer 3: ```suggestion
When a controller receives discovery response with a federated service, it creates `ServiceEntry` or `WorkloadEntry`
```

> in the local cluster depending on whether the service exists locally.

Author: I changed that section, please read again. :)

The controller also watches local `Service` objects and checks if they match the export rules - if so,

> it adds the service's FQDN to the auto-passthrough `Gateway` hosts.

Reviewer 3: ```suggestion
it adds the service's FQDN to the auto-passthrough `Gateway` hosts, allowing them to be discovered by remote federated meshes.
```
Again, not sure if I have the above quite right - just want a bit more explanation to ensure the point is understood.

### Architecture

> ![architecture](docs/img/architecture.jpg)

Reviewer 3: I notice there is only a Gateway shown on the east side... is that for simplicity in the diagram, or is it possible that there is only a gateway on the east side?

Author: Yes, it is. I changed the diagram.
